File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    boolean isCan(int[] weights , int days , int mid){
        int d = 1;
        int sum = 0;
        for(int x : weights){
            if(sum + x <= mid){
                sum += x;
                continue;
            }else{
                d++;
                sum = x;
            }
        }
        return d <= days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int n = weights.length;
        int l = Arrays.stream(weights).max().getAsInt();
        int r = Arrays.stream(weights).sum();
        int ans = Integer.MAX_VALUE;
        while(l <= r){
            int mid = l + (r - l) / 2;
            if(isCan(weights , days, mid)){
                ans = Math.min(ans , mid);
                r = mid - 1;
            }else {
                l = mid + 1;
            }
        }
        return ans;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool isCan(vector<int>& weights , int days , int mid){
        int n = weights.size();
        int d = 1;
        int sum = 0;
        for(int i = 0 ;  i < n ;i++){
            if(sum + weights[i] <= mid){
                sum += weights[i];
                continue;
            }
            else{
                d++;
                sum = weights[i];
            }
        }
        return d <= days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int n = weights.size();
        int l = *max_element(begin(weights) , end(weights));
        int r = accumulate(begin(weights) , end(weights) , 0);
        int ans = INT_MAX;
        while(l <= r){
            int mid = (l + r) / 2;
            if(isCan(weights , days , mid)){
                ans = min(ans , mid);
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return ans;
    }
}; 